### PR TITLE
Align `flutter pub get/upgrade/add/remove/downgrade`

### LINKED ISSUE #3717: flutter pub advertizes an option that doesn't exist
```
$ flutter pub get
Resolving dependencies in ./example failed. For details run `flutter pub get --directory ./example`
pub get failed
command: "/Users/mit/dev/flutter/bin/cache/dart-sdk/bin/dart __deprecated_pub --color --directory . get --example"
pub env: {
  "FLUTTER_ROOT": "/Users/mit/dev/flutter",
  "PUB_ENVIRONMENT": "flutter_cli:get",
  "PUB_CACHE": "/Users/mit/.pub-cache",
}
exit code: 1

$ flutter pub get --directory ./example/
Could not find an option named "directory".
```

### PR DESCRIPTION
**Pull request by sigurdm**
This PR does some tidying of the `flutter pub get/upgrade/add/remove/downgrade` command pipelines

* Now all of these commands are run in ProcessMode.inheritStdio (before only `get` was).
* Now `get` can be invoked with a '--directory' argument. (https://github.com/dart-lang/pub/issues/3717)
  *  Specifying the target directory as a bare argument is deprecated with a warning on the console.
* Now `get --help` will forward output from `dart pub get --help`
* All of these commands will send usage metrics and timings (before only `get` was doing this)

This PR also removes the `Running "flutter pub get" in <direcotry>...` line, as this is redundant with pubs own Resolving dependencies in <>...

  Before:
  ```
  > flutter pub get
  Running "flutter pub get" in flutter_tools...
  Resolving dependencies... (2.2s)
    _fe_analyzer_shared 51.0.0 (52.0.0 available)
    analyzer 5.3.1 (5.4.0 available)
    archive 3.3.2 (3.3.5 available)
    checked_yaml 2.0.1 (2.0.2 available)
    intl 0.17.0 (0.18.0 available)
    native_stack_traces 0.5.2 (0.5.5 available)
  Got dependencies!
  > flutter pub get ../flutter_test
  Running "flutter pub get" in flutter_test...
  Resolving dependencies in ../flutter_test... 
  Got dependencies in ../flutter_test!
  ```

  After:
  ```
  > flutter pub get
  Resolving dependencies... (2.1s)
    _fe_analyzer_shared 51.0.0 (52.0.0 available)
    analyzer 5.3.1 (5.4.0 available)
    archive 3.3.2 (3.3.5 available)
    checked_yaml 2.0.1 (2.0.2 available)
    intl 0.17.0 (0.18.0 available)
    native_stack_traces 0.5.2 (0.5.5 available)
  Got dependencies!
  > flutter pub get ../flutter_test
    Using a naked argument for directory is deprecated and will stop working in a future Flutter release.
  
    Use --directory instead.
  Resolving dependencies in ../flutter_test... 
  Got dependencies in ../flutter_test!
  > flutter pub get --directory=../flutter_test
  Resolving dependencies in ../flutter_test... 
  Got dependencies in ../flutter_test!
  ```

Fixes https://github.com/dart-lang/pub/issues/3717

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

### CODE REVIEW

**Comment by sigurdm** [comment]
There are still a bunch of tests that could be written to validate new behavior - but I would like feedback on the change before digging too deep.

**Review by jonasfj — APPROVED** [review]

**Comment by jonasfj** [comment]
Obviously, we need to fix the tests. But from my quick look this hopefully won't break too much.

File: packages/flutter_tools/lib/src/dart/pub.dart [hunk 1]
```diff
@@ -743,6 +687,32 @@ class _DefaultPub implements Pub {
...
+    if (!project.packageConfigFile.existsSync()) {
+      throwToolExit('${project.directory}: pub did not create .dart_tools/package_config.json file.');
+    }
+    final File lastVersion = _fileSystem.file(
+      _fileSystem.path.join(project.directory.path, '.dart_tool', 'version'),
+    );
+    final File currentVersion = _fileSystem.file(
+      _fileSystem.path.join(Cache.flutterRoot!, 'version'));
+      lastVersion.writeAsStringSync(currentVersion.readAsStringSync());
+    lastVersion.writeAsStringSync(currentVersion.readAsStringSync());
```
Comment (nehalvpatel) [inline packages/flutter_tools/lib/src/dart/pub.dart:708]: Seems like a typo here with this line being duplicated.

**Review by nehalvpatel — APPROVED** [review]
LGTM 👍

File: packages/flutter_tools/lib/src/dart/pub.dart [hunk 2]
```diff
@@ -755,11 +725,8 @@ class _DefaultPub implements Pub {
   ///
   /// For more information, see:
   ///   * [generateLocalizations], `in lib/src/localizations/gen_l10n.dart`
-  Future<void> _updatePackageConfig(
-    File packageConfigFile,
-    Directory generatedDirectory,
-    bool generateSyntheticPackage,
```
Comment (nehalvpatel) [inline packages/flutter_tools/lib/src/dart/pub.dart:761]: Nit: seems like the doc for this method is outdated now with the reference to the `generateSyntheticPackage` argument

**Comment by christopherfujino** [comment]
Looks like this has legitimate internal Google testing failures that will require a g3fix: cl/500157954

**Comment by mit-mit** [comment]
@christopherfujino can we go ahead and merge this, or do we need to resolve the internal testing issue first?

**Review by christopherfujino — APPROVED** [review]
Ship it (Google testing is green)

**Comment by christopherfujino** [comment]
I added `autosubmit` label, so the bot should merge it once the tree goes green.

**Pull request merged**